crates/dojo-core/src/database.cairo

```diff
@@ -33,48 +56,121 @@ fn set(table: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: S
     storage::set_many(0, keys.span(), offset, value, layout);
 }
 
-fn set_with_index(
-    table: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>
+/// Creates an entry in the database and adds it to appropriate indexes.
+/// # Arguments
+/// * `table` - The table to create the entry in.
+/// * `id` - id of the created entry.
+/// * `keys` - The keys to index the entry by.
+/// * `offset` - The offset of the entry.
+/// * `value` - The value of the entry.
+/// * `layout` - The layout of the entry.
+fn set_with_index( 
+    table: felt252,
+    id: felt252,
+    keys: Span<felt252>,
+    offset: u8,
+    value: Span<felt252>,
+    layout: Span<u8>
 ) {
-    set(table, key, offset, value, layout);
-    index::create(0, table, key);
+    set(table, id, offset, value, layout);
+    index::create(0, table, id, 0); // create a record in index of all records
+
+    let mut idx = 0;
+    loop {
+        if idx == keys.len() {
+            break;
+        }
+        let mut serialized = ArrayTrait::new();
+        table.serialize(ref serialized);
+        idx.serialize(ref serialized);
+        let index = poseidon_hash_span(serialized.span());
+
+        index::create(0, index, id, *keys.at(idx)); // create a record for each of the keys
+        
+        idx += 1;
+    };
+
+    let len_keys = array!['dojo_storage_keys_len', table, id].span();
+    storage::set(0, len_keys, keys.len().into()); // save the number of keys
```

Comment:
hmm is this the same for every set of a model, assuming the indexed keys are constant?

Reply:
Yes, as long as keys don't change , and are of constant size, this will be equal. If we assume that, this `len` can be passed to `del`, without a need for storage at all